src/net/server/mod.rs

```diff
@@ -0,0 +1,245 @@
+use std::net::SocketAddr;
+
+use axum::{
+    extract::rejection::QueryRejection,
+    response::{IntoResponse, Response},
+    routing::get,
+    Router,
+};
+use axum_server::{tls_rustls::RustlsConfig, Handle};
+use hyper::StatusCode;
+use thiserror::Error;
+use tokio::task::JoinHandle;
+
+mod handlers;
+
+#[derive(Error, Debug)]
+pub enum MpcServerError {
+    #[error(transparent)]
+    BadQueryString(#[from] QueryRejection),
+}
+
+impl IntoResponse for MpcServerError {
+    fn into_response(self) -> Response {
+        let status_code = match &self {
+            MpcServerError::BadQueryString(_) => StatusCode::BAD_REQUEST,
+        };
+
+        (status_code, self.to_string()).into_response()
+    }
+}
+
+/// Axum router definition for MPC helper endpoint
+#[must_use]
+pub fn router() -> Router {
+    Router::new().route("/echo", get(handlers::echo_handler))
+}
+
+/// MPC helper supports HTTP and HTTPS protocols. Only the latter is suitable for production,
+/// http mode may be useful to debug network communication on dev machines
+pub enum BindTarget {
+    Http(SocketAddr),
+    Https(SocketAddr, RustlsConfig),
+}
+
+/// Starts a new instance of MPC helper and binds it to a given target.
+/// Returns a socket it is listening to and the join handle of the web server running.
+pub async fn bind(target: BindTarget) -> (SocketAddr, JoinHandle<()>) {
+    let svc = router().into_make_service();
+    let handle = Handle::new();
+
+    let task_handle = match target {
+        BindTarget::Http(addr) => tokio::spawn({
+            let handle = handle.clone();
+            async move {
+                axum_server::bind(addr)
+                    .handle(handle)
+                    .serve(svc)
+                    .await
+                    .expect("Failed to serve");
+            }
+        }),
+        BindTarget::Https(addr, tls_config) => tokio::spawn({
+            let handle = handle.clone();
+            async move {
+                axum_server::bind_rustls(addr, tls_config)
+                    .handle(handle)
+                    .serve(svc)
+                    .await
+                    .expect("Failed to serve");
+            }
+        }),
+    };
+
+    (
+        handle
```

Comment:
what does it mean for you to return `handle` and `task_handle` (which contains a `handle.clone()`) here?

Reply:
I return `SocketAddr` here in case if clients don't want to specify a port to bind to, in this case Kernel pick it up for us, `Handle` type is here to be able to get this information (it only becomes available after Kernel signals us which port is assigned). `JoinHandle` is the Tokio task handle that runs the HTTP(s) server, clients may want to abort it or (like CLI) just await it until user terminates it

Reply:
ok, think I got it.

so the `.listening()` function is async. i.e. it returns when listening has started, not ended.

what do you expect a caller of `bind` to do with the returned `SocketAddr`?

Reply:
they should advertise it as the address they're listening to - for example, in unit tests I don't want to specify a port to bind to avoid "address already in use" error and let kernel to find any port that is available. After that, I can connect to this address using an HTTP client and make requests

Reply:
ok, gotcha. so mostly if you are going to subsequently connect a client to that server which has a randomly assigned port